src/com/trilead/ssh2/crypto/dh/DhExchange.java

```diff
@@ -1,160 +1,209 @@
+			"4009438B481C6CD7889A002ED5EE382BC9190DA6FC026E47" +
+			"9558E4475677E9AA9E3050E2765694DFC81F56E880B96E71" +
+			"60C980DD98EDD3DFFFFFFFFFFFFFFFFF", 16);
+
+	private static final BigInteger G = BigInteger.valueOf(2);
+
+	/* Hash algorithm to use */
+	private String hashAlgo;
 
 	/* Client public and private */
 
-	BigInteger e;
-	BigInteger x;
+	private DHPrivateKey clientPrivate;
+	private DHPublicKey clientPublic;
 
 	/* Server public */
 
-	BigInteger f;
-
-	/* Shared secret */
-
-	BigInteger k;
-
-	static
-	{
-		final String p1_string = "17976931348623159077083915679378745319786029604875"
-				+ "60117064444236841971802161585193689478337958649255415021805654859805036464"
-				+ "40548199239100050792877003355816639229553136239076508735759914822574862575"
-				+ "00742530207744771258955095793777842444242661733472762929938766870920560605"
-				+ "0270810842907692932019128194467627007";
-
-		final String p14_string = "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD129"
-				+ "024E088A67CC74020BBEA63B139B22514A08798E3404DDEF9519B3CD3A431B302B0"
-				+ "A6DF25F14374FE1356D6D51C245E485B576625E7EC6F44C42E9A637ED6B0BFF5CB"
-				+ "6F406B7EDEE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3DC2007CB8A"
-				+ "163BF0598DA48361C55D39A69163FA8FD24CF5F83655D23DCA3AD961C62F356208"
-				+ "552BB9ED529077096966D670C354E4ABC9804F1746C08CA18217C32905E462E36C"
-				+ "E3BE39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9DE2BCBF69558171"
-				+ "83995497CEA956AE515D2261898FA051015728E5A8AACAA68FFFFFFFFFFFFFFFF";
-
-		p1 = new BigInteger(p1_string);
-		p14 = new BigInteger(p14_string, 16);
-		g = new BigInteger("2");
-	}
-
-	private final String hashAlgorithm;
-
-	@Deprecated
-	public DhExchange()
-	{
-		this("SHA1");
-	}
-
-	public DhExchange(String hashAlgorightm) {
-		super();
-		this.hashAlgorithm = hashAlgorightm;
-	}
-
-	public void init(int group, SecureRandom rnd)
-	{
-		k = null;
-
-		if (group == 1)
-			p = p1;
-		else if (group == 14)
-			p = p14;
-		else
-			throw new IllegalArgumentException("Unknown DH group " + group);
-
-		x = new BigInteger(p.bitLength() - 1, rnd);
-
-		e = g.modPow(x, p);
-	}
-
-	/**
-	 * @return Returns the e.
-	 *
-	 */
-	public BigInteger getE()
-	{
-		if (e == null)
-			throw new IllegalStateException("DhDsaExchange not initialized!");
+	private DHPublicKey serverPublic;
+
+	@Override
+	public void init(String name) throws IOException {
+		final DHParameterSpec spec;
+		if ("diffie-hellman-group18-sha512".equals(name)) {
+			spec = new DHParameterSpec(P18, G);
+			hashAlgo = "SHA-512";
+		} else if ("diffie-hellman-group16-sha512".equals(name)) {
+			spec = new DHParameterSpec(P16, G);
+			hashAlgo = "SHA-512";
+		} else if ("diffie-hellman-group14-sha256".equals(name)) {
+			spec = new DHParameterSpec(P14, G);
+			hashAlgo = "SHA-256";
+		} else if ("diffie-hellman-group14-sha1".equals(name)) {
+			spec = new DHParameterSpec(P14, G);
+			hashAlgo = "SHA-1";
+		} else if ("diffie-hellman-group1-sha1".equals(name)) {
```

Comment:
Another algorithm to remove while hardening.

Reply:
@jvz should I remove support in the library in this commit? Do you mean that I am introducing an algorithm that has already been removed?

Reply:
I think its strange to remove kex algorithms from a ssh client library because of hardening. 
(especially since all the preexisting supported one are all weaker then all of the new ones added!)

the right way to do the hardening would be to at the usage of. 
com.trilead.ssh2.Connection configure the cryptoWishList used to establish the connection. 
its already possible to alter the wishlist for all other 
Connection::setClient2ServerCiphers
Connection::setClient2ServerMACs
Connection::setServer2ClientCiphers
Connection::setServer2ClientMACs
Connection::setServerHostKeyAlgorithms
but not the kex for some reason.

Reply:
Sure, but key exchange is the weakest link in an SSH connection ;)

Reply:
caring people would ofcourse configure their trilead client something like this:
Connection::setKexAlgorithms("curve25519-sha256","ecdh-sha2-nistp521","ecdh-sha2-nistp384","ecdh-sha2-nistp256");
(which would mean that it ignores all supported "diffie-hellman")

What really meant it a bit strange to disagree with a pullrequest that is only added KEX that are all more secure than the existing supported ones because some of the added are not secure enough.(even through they are better than the current master branch) 

Currently trilead only supports weak kex: 
diffie-hellman-group-exchange-sha256, diffie-hellman-group-exchange-sha1, diffie-hellman-group14-sha1, diffie-hellman-group1-sha1

Reply:
This PR has been recreated in https://github.com/jenkinsci/trilead-ssh2/pull/60 to handle the key exchange additions. This original PR practically rewrites the library. And all those SHA-1 exchanges are deprecated. Are you using this library outside Jenkins? Because if so, you might be better off using Apache Mina or sshj or something up to date.